[SingleStore] Add SingleStore connector

mitchwadair opened the pull request:
This PR intends to add SingleStore support to Drizzle ORM. SingleStore is a distributed SQL database that unifies both transactional and analytical workloads. It is largely compatible with MySQL, but there are some incompatibilities and additional features that are supported.

## Key differences between MySQL vs SIngleStore Integration Tests

- Renamed all references from MySQL to SingleStore in the integration tests directory.
- The tests are performed against the [singlestore-dev](https://github.com/singlestore-labs/singlestoredb-dev-image) docker image, to replicate a production environment
- SingleStore's `serial` column type only assures uniqueness of the column values. The tests were modified to (set operation tests in particular) to expect SingleStore-specific behavior, by adding an `ORDER BY` instruction to every select instruction.
- `ORDER BY` and `LIMIT` cannot be chained together.
- Removed SingleStore proxy package and its IT tests. If it's necessary, this package can be replicated from the MySQLL proxy package later
- The foreign key integration tests were removed since they are not supported by singlestore
- Removed `intersectAll` and `exceptAll` tests since these operations are not supported by SIngleStore
- Removed nested transaction tests since this feature is not supported by [SingleStore](https://docs.singlestore.com/cloud/reference/sql-reference/procedural-sql-reference/transactions-in-stored-procedures/)
- Removed `isolationLevel` config and tests as singlestore [only supports](https://docs.singlestore.com/cloud/getting-started-with-singlestore-helios/about-singlestore-helios/singlestore-helios-faqs/durability/) one `isolationLevel`

### Tests results for SingleStore

`Drizzle-kit` tests result

![image](https://github.com/user-attachments/assets/8d0705b1-3972-4256-b6d9-2b6165511bc7)

`Drizzle-orm` tests result

![image](https://github.com/user-attachments/assets/45f15a76-faca-46dc-8f44-90651ba56361)

`Drizzle-orm` type tests result

![image](https://github.com/user-attachments/assets/84f16da1-4c44-4572-b015-2bcf4bfb5d81)

`Integration-tests` results

![image](https://github.com/user-attachments/assets/4022ac8c-4c22-4d14-9f5e-2a5b607f855f)

## Side notes
There are some SingleStore-specific operations, such as `attach/detach`, `create/drop milestone` and `optimizeTable` for column stores tables, that are implemented on `drizzle-orm/src/singlestore-core` but not on `drizzle-kit`. There are also no integration tests for these operations. This is a known issue and should be fixed in the upcoming releases.

The relational tests that included nested selects with aggregations functions were skipped since SingleStore doesn't support it

Rodriguespn committed:
[SingleStore] Add SingleStore connector (#32)

Rodriguespn committed:
[SingleStore] Add SingleStore connector (#32)

Commit:
rebase singlestore connector with last commit: https://github.com/drizzle-team/drizzle-orm/commit/526996bd2ea20d5b1a0d65e743b47e23329d441c

Commit:
rebase singlestore connector with last commit: https://github.com/drizzle-team/drizzle-orm/commit/526996bd2ea20d5b1a0d65e743b47e23329d441c

Commit:
skip view tests

Commit:
skip view tests

AlexBlokh committed:
imports checker with OHM grammar

AndriiSherman commented:
Starting my review today

AndriiSherman commented:
I'll resolve conflicts while doing my review, so no action is needed from your side

AlexBlokh committed:
imports checker with OHM grammar + fix of imports issues

AndriiSherman committed:
Merge branch 'main' of https://github.com/drizzle-team/drizzle-orm into feature/new-dialect-singlestore

AndriiSherman committed:
Merge branch 'beta' of https://github.com/drizzle-team/drizzle-orm into feature/new-dialect-singlestore

AndriiSherman committed:
Fix conflicts

AndriiSherman committed:
Fix conflicts

AndriiSherman committed:
No fks?

AndriiSherman committed:
No fks?

AndriiSherman committed:
Fix after failed tests

AndriiSherman committed:
Fix after failed tests

apeng-singlestore committed:
Add drizzleIntegration query header

Added query header for future logging of integration usage

apeng-singlestore committed:
Add drizzleIntegration query header

Added query header for future logging of integration usage

AndriiSherman committed:
Add a few test changes

Using database for tests and remove datetime and json from singlestore allTypes test on push

AndriiSherman committed:
Add a few test changes

Using database for tests and remove datetime and json from singlestore allTypes test on push

Commit:
fix some singlestore drizzle kit tests

apeng-singlestore committed:
Merge branch 'feature/new-dialect-singlestore' of https://github.com/singlestore-labs/s2-drizzle-orm-dev into feature/new-dialect-singlestore

AndriiSherman committed:
Merge branch 'feature/new-dialect-singlestore' of github.com:singlestore-labs/s2-drizzle-orm-dev into feature/new-dialect-singlestore

Commit:
update SingleStore decimal column type

Commit:
remove unnecessary logs

AndriiSherman committed:
Fixed type parsing

AndriiSherman committed:
Merge branch 'feature/new-dialect-singlestore' of github.com:singlestore-labs/s2-drizzle-orm-dev into feature/new-dialect-singlestore

AndriiSherman committed:
Fix decimal SingleStore

Commit:
update SingleStore decimal column type

Commit:
update SingleStore float column type

Commit:
singlestore drizzle-kit test pass

Commit:
nit: missing files from main

Commit:
fix changelog files

Commit:
fix drizzle-kit non singlestore files

Commit:
remain fix of non singlestore files and common files

Commit:
update drizzle-orm

Commit:
update singlestore-driver

Commit:
update singlestore-proxy

Commit:
cleaning singlestore drizzle-orm and passing tests

Commit:
remove views and generated columns from drizzle-kit

Commit:
remove mode from drizzle-orm

Commit:
fix replica integration tests

AndriiSherman committed:
Fixes for kit singlestore

AndriiSherman committed:
Remove unsupported API

AndriiSherman committed:
Merge branch 'beta' of https://github.com/drizzle-team/drizzle-orm into feature/new-dialect-singlestore

AndriiSherman committed:
Fix build problems

AndriiSherman committed:
Clear drizzle-kit tests

AndriiSherman committed:
Remove .only from tests

AndriiSherman committed:
Merge branch 'beta' of https://github.com/drizzle-team/drizzle-orm into feature/new-dialect-singlestore

AndriiSherman committed:
Fix integration tests by removing RQB support for now

AndriiSherman committed:
Fix tests
